text/0000-process-specific-env.md

```diff
@@ -0,0 +1,67 @@
+- CNB Issue: 
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Each process type could have its own independent environment variables.
+
+# Motivation
+[motivation]: #motivation
+
+- Greater flexibility to define runtime processes
+- When a new process type is created it often has different requirements for environment variables
+
+One example is a buildpack requiring different `NODE_OPTIONS` for a `web` process and a `debug` process. This would allow a node buildpack author to write `<layer>/env.launch/debug/NODE_OPTIONS` with a value of `--inspect=0.0.0.0:9229` for `debug` while omitting the file entirely for `web`. While this is currently possible for the buildpack author when defining the launch processes, this change would also allow ad-hoc `launcher` commands to ignore `envs` intended for specific processes.
+
+```
+# defined processes
+/cnb/lifecycle/launcher debug # NODE_OPTIONS=--inspect=0.0.0.0:9229
+/cnb/lifecycle/launcher web # no NODE_OPTIONS defined
+
+# ad-hoc
+/cnb/lifecycle/launcher -- npm install # no NODE_OPTIONS defined
+/cnb/lifecycle/launcher -- bash # no NODE_OPTIONS defined
+```
+
+`<layer>/profile.d/debug/debug.sh` would work similarly and be sourced by Bash before launching the `debug` command, but not when executing any other process.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Each env or profile layer dir can have subdirs like `<layer>/env.launch/web/` and `<layer>/profile.d/web/`. These behave the same way as `<layer>/env` but are specific to the launch of the `web` process.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Lifecycle would use the identifier of the intended launch process to bring in additional `env` and `profile.d` files. The launcher would process these process specific values _after_ less specific values to allow [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/master/buildpack.md#environment-variable-modification-rules) to apply in the context of a single buildpack.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Additional complexity would be introduced into the launcher.
+- Introducing process specific directory structure for `env` and `profile.d` but not `bin` and `lib`, for example, could be confusing for buildpack authors.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do Nothing. Buildpack authors can continue work around some of the restrictions today by embedding process specific `env` values into the launch processes instead of using the `env` directory.
+
+# Prior Art
+[prior-art]: #prior-art
+
+- Copied from [here](https://github.com/buildpacks/rfcs/pull/72) and updated to include additional details and examples.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should this idea be extended to `bin`, `lib`, and any other buildpack output intended for launch?
+- Should lifecycle ignore invalid entries like `<layers>/<layer>/env.build/web/NODE_OPTIONS`?

```

Comment:
I'm failing to see what is invalid in this entry...

Reply:
For this entry it specifies build and a process. Since env vars would be process specific this wouldn’t make since as you aren’t building a specific process.

Reply:
Ahh, I see now.